src/Be.Vlaanderen.Basisregisters.Projector/Internal/Runners/ConnectedProjectionsCatchUpRunner.cs

```diff
@@ -13,34 +14,83 @@ internal class ConnectedProjectionsCatchUpRunner
         private readonly IReadonlyStreamStore _streamStore;
-        private readonly IConnectedProjectionEventBus _eventBus;
+        private readonly IProjectionManager _projectionManager;
         private readonly ILogger<ConnectedProjectionsCatchUpRunner> _logger;
 
         public ConnectedProjectionsCatchUpRunner(
             IReadonlyStreamStore streamStore,
             ILoggerFactory loggerFactory,
-            IConnectedProjectionEventBus eventBus,
-            IConnectedProjectionEventHandler connectedProjectionEventHandler)
+            IProjectionManager projectionManager)
         {
             _projectionCatchUps = new Dictionary<ConnectedProjectionName, CancellationTokenSource>();
             _streamStore = streamStore ?? throw new ArgumentNullException(nameof(streamStore));
-            _eventBus = eventBus ?? throw new ArgumentNullException(nameof(eventBus));
+            _projectionManager = projectionManager ?? throw new ArgumentNullException(nameof(projectionManager));
             _logger = loggerFactory?.CreateLogger<ConnectedProjectionsCatchUpRunner>() ?? throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        public bool IsCatchingUp(ConnectedProjectionName projectionName)
+        {
+            return projectionName != null &&
+                   _projectionCatchUps.ContainsKey(projectionName);
+        }
+
+        public void HandleCatchUpCommand<TCatchUpCommand>(TCatchUpCommand command)
+            where TCatchUpCommand : CatchUpCommand
+        {
+            _logger.LogTrace("CatchUp: Handling {Command}", command);
+            switch (command)
+            {
+                case StartCatchUp startCatchUp:
+                    Handle(startCatchUp);
+                    break;
+                case RemoveStoppedCatchUp removeStoppedCatchUp:
+                    Handle(removeStoppedCatchUp);
+                    break;
+                case StopCatchUp stopCatchUp:
+                    Handle(stopCatchUp);
+                    break;
+                case StopAllCatchUps _:
+                    StopAllCatchUps();
+                    break;
+                default:
+                    _logger.LogError("No handler defined for {Command}", command);
+                    break;
+            }
+        }
 
-            if (connectedProjectionEventHandler == null)
-                throw new ArgumentNullException(nameof(connectedProjectionEventHandler));
+        private void Handle(StopCatchUp stopCatchUp)
+        {
+            if (stopCatchUp.ProjectionName == null || false == IsCatchingUp(stopCatchUp.ProjectionName))
+                return;
 
-            connectedProjectionEventHandler
-                .RegisterHandleFor<CatchUpStopped>(message => _projectionCatchUps.Remove(message.Projection));
+            try
+            {
+                var catchUp = _projectionCatchUps[stopCatchUp.ProjectionName];
+                if (false == catchUp.IsCancellationRequested)
```

Comment:
this is hard to read, does it do `catchUp.Cancel` when `IsCancellationRequested` is false? Can you flip this and do `if (catchup.IsCancellationRequest)` it reads nicer and I think if fixes a bug in this case?

Reply:
What's the (potential) bug you saw here?

Reply:
So the original code reads:

```csharp
if (false == catchUp.IsCancellationRequested)
    catchUp.Cancel();```

If we flip the operator for readability we get:

```csharp
if (catchUp.IsCancellationRequested == false)
    catchUp.Cancel();```

Or

```csharp
if (!catchUp.IsCancellationRequested)
    catchUp.Cancel();```

Which I read as:
> If NOT cancel requested, DO cancel

So without knowing in great detail how it works, I thought this is either a potential bug, or a naming issue, or my misunderstanding. You aren't asking for a cancellation, but it gets cancelled?

Reply:
Was more a safeguard to not do a double cancellation, but that's caught in the Cancel() method anyway. (just checked the implementation)
Since that check is not really needed changed the complete block to:

```csharp
            try
            {
                using (var catchUp = _projectionCatchUps[projectionName])
                    catchUp.Cancel();
            }
            catch (KeyNotFoundException) { }
            catch (ObjectDisposedException) { }
```